src/Control/Lens/Internal/Equality.hs

```diff
@@ -0,0 +1,83 @@
+{-# language CPP #-}
+{-# language GADTs #-}
+{-# language RankNTypes #-}
+{-# language StandaloneDeriving #-}
+{-# language TypeOperators #-}
+module Control.Lens.Internal.Equality (
+    (:~:)(..)
+  , sym
+  , trans
+  , castWith
+  , gcastWith
+  , apply
+  , inner
+  , outer
+  , eqT
+  ) where
+#if MIN_VERSION_base (4,7,0)
+import Data.Type.Equality
+import Data.Typeable (eqT)
+
+#else
+
+import Control.Category
+import Data.Typeable (Typeable, gcast)
+import Prelude hiding (id, (.))
+
+infix 4 :~:
+
+-- | Propositional equality. If @a :~: b@ is inhabited by some terminating
+-- value, then the type @a@ is the same as the type @b@. To use this equality
+-- in practice, pattern-match on the @a :~: b@ to get out the @Refl@ constructor;
+-- in the body of the pattern-match, the compiler knows that @a ~ b@.
+data a :~: b where
+  Refl :: a :~: a
+
+instance Category (:~:) where
+  id = Refl
+  Refl . Refl = Refl
+
+-- | Symmetry of equality
+sym :: (a :~: b) -> (b :~: a)
+sym Refl = Refl
+
+-- | Transitivity of equality
+trans :: (a :~: b) -> (b :~: c) -> (a :~: c)
+trans Refl Refl = Refl
+
+-- | Type-safe cast, using propositional equality
+castWith :: (a :~: b) -> a -> b
+castWith Refl x = x
+
+-- | Generalized form of type-safe cast using propositional equality
+gcastWith :: (a :~: b) -> ((a ~ b) => r) -> r
+gcastWith Refl x = x
+
+-- | Apply one equality to another, respectively
+apply :: (f :~: g) -> (a :~: b) -> (f a :~: g b)
+apply Refl Refl = Refl
+
+-- | Extract equality of the arguments from an equality of applied types
+inner :: (f a :~: g b) -> (a :~: b)
```

Comment:
Interestingly, GHC 7.4 and 7.6 appear to be unable to typecheck `inner` (again, likely due to old GHC bugs). Let's just leave out `inner`.

Reply:
In theory, you should be able to convert the equality to a Liebnizian equality and use the approach I use in the old `eq` package in order to get `inner` to work, as that seems to work from 7.0 on, if we want it.